README.md
# logicTester

This is a POC (proof-of-concept) of a simple but configurable 'SSI'-logic I.C-tester.

Unlike some other testers, you have to specify which I.C. you have and the script
reports proper working or a failure. 

This tester splits the low-level measuring tasks from the high-level data 
library, test data generation and comparison tasks. The low-level functionality 
is implemented in an Arduino sketch which communicates with the high-level control 
and compare functionality build in Python. The Python script does contain the I.C. 
data, this is read from a text-based configuration file. As long as the behaviour 
at the output pins can be defined in Boolean logic, the chip can be tested. 

The current version is limited to stateless gates and inverters. So far I didn't 
try to define a machine-parsable notation for I.C.s with clocked or register
functionality. A limitation of the Arduino sketch is that it currently only 
supports 14 and 16 pin I.C.s and the switch is manual. 

Only the logic behaviour is tested, not wether the I.C. conforms to TTL or CMOS 
specifications.

## Hardware

![Arduino Uno with cheap shield and miniature experimenter board, complete with test I.C.; a working N8T97N](thirdPrototype.jpg)

![power switching circuit for the 14-pin I.C. The same pin is used as signal pin in 16-pin mode.](pin14power.png)

The latest version of the prototype includes a software switch for the power pin 
of 14-pin I.C.s. I tend to forget to switch the power pin, resulting in much
unneeded debugging sessions. The same pin is used for a signal in 16-pin mode, 
requiring a MosFet to remove power. The 16-pin I.C. power is still always on.

The planned PCB-version will have power switching for 'all' possible power pins. 

## Configuration

The configuration file format is currently implemented as json, but omitting the outer braces.

    "type": "7402", "pins": 14, "config": "C:Q,2,3,Q,5,6,G,8,9,Q,11,12,Q,V",  "M1": "!(2|3)", "M4": "!(5|6)", "M10": "!(8|9)",  "M13":  "!(11|12)"

In the config section, the inputs (exercise pins) are numbered as the pin number, 
the outputs indicated as 'Q', the ground and Vcc pins as 'G' and 'V'. This is a 
compromise between human-readable, machine readable and compactness. 

The M? sections relate to the I.C output pins and contain a Boolean-expression
defining the output in terms of input lines and characters defining logical
expressions; ! = NOT, | is OR, & is AND and ^ as XOR.

## Usage

The Python script is written with Python 3.6.9 on Linux Xubuntu 18.04.

The script is started with ***python3 &lt;port&gt; &lt;chip&gt;***

***port*** the the serial port, Baud rate is 9600 Bd.

***chip*** is the chip to test. Currently only the 74(LS)00 and 74(LS)02 are 
completely implemented.

Example:

    python3 icTest.py /dev/ttyUSB1 7402

## Message format

This is the current help message from the Arduino:

    ICtest 1.3
    C - configure pins
    D - debug mode
    E - exercise pin with 500ms cycle
    H - this text
    Q - set and query pins
    R - reset config and pins


### To Arduino:

* **C:&lt;pin-spec&gt;** - configure the pins on the Arduino. If the pin-numbering is
supported the response is "**OK**". If not the response is "**ERROR**".

Example: 

    C:Q,2,3,Q,5,6,G,8,9,Q,11,12,Q,V

* **Q:&lt;set-and-query&gt;** - specifies the values for the I.C. input pins. The 
response is modified copy;the first char is '**R**' and the I.C. output pin values 
are filled in.

Example: 

    Q:-,1,0,-,0,0,G,0,0,-,0,0,-,V

* **R** - reset all pins to tri-state and erase configuration.

### From Arduino:

* OK - response to **C:&lt;pin-spec&gt;** and **R**

* **R:&lt;query_result&gt;** similar to the **Q:-,1,0,-,0,0,G,0,0,-,0,0,-,V**
but with output pin levels filled in. Using "H" and "L" so the difference between
input and output pins is still visible.

Example:
 
    R:L,0,1,1,L,0,G,0,0,H,0,0,H,V

### Stand alone Arduino commands
    
The Arduino code has some extra commands not used by the Python code, but useful
for 'manual' debugging.

* **D** - switch on and off debug mode. With debug on the Arduino becomes more 
talkative, but this is not supported by the Python script. De arguments are 1 for 
on and 0 for off. One character has to be between the 'D' and the number.

Example:

    D:1
    
* **E** - this generates a 500 ms square wave on one pin. The idea is to set up
a pattern of 1's and 0's on all input pins using the **Q:** command and use **E**
to see the response on the output pins. The default check pattern from the Python
script isn't very intuitive for this.

Example:

    E:11
    



## Process

The Python script sends the config string to the Arduino, which initialises the
pins. It then calculates the number of permutations and sends a query for each 
to the Arduino. It also calculates the expected response based on the Boolean
expressions in the 'M? sections.

The response of the Arduino is compared with this expected response and 
differences are flagged. On completion the number of errors is reported. Default 
the script is quite chatty, so you can see what it does. Even more print's are 
commented out, very handy for debugging. The Arduino sketch cost me one day, the 
Python script two, almost debugging each individual line with print()'s.

Also see the [ToDo](ToDo.txt) for what isn't there yet but could be.

There is also a <a href="https://hackaday.io/project/175252-simple-ttl-logic-tester">Hackaday page</a>.


ReleaseNotes.txt
Added to version 1.3 (2020-10-10):

Arduino code:
- Exercise command E:p (p is the I.C. pin number) for one pin (500ms cycle)
- Power switching for 14-pin I.C.s with a P-channel MosFet
- Debug command D:n (1 = on, 0 is off)


Fixed in version 1.2 (2020-10-08):

- Ignored NC pins.
  Each unused exercise pin doubles the number of mutations, for the 7430 the 
  difference is 2048 : 256, a factor 8
- Tested actual XOR function: 7486 working
- Added more I.C.s
- Converted config file format to full JSON
- Fixed 16 pin mapping
- Changed the response for query pins from 0/1 to H/L making the output more readable, 
  very handy for debugging new boolean logic definitions
- Changed the Baud rate, as 1024 and up permutations takes quite a while.


Fixed in version 1.1 (2020-10-05):
- Add more I.C. definitions and test them


icTest.py
#!/usr/bin/python3
#

import sys
import serial 
import time
import re
import json 

VERSION = "1.3"
    
ttySpeed = 115200
ttyPort = ""
icType = ""
libraryName = 'icLibrary.txt'

arduinoWait = 2
DEBUG = False

def log(message):
    if (DEBUG):
        print(message)

def writeSerial(port, message):
    port.write (str.encode(message + "\r\n"))
#    time.sleep(arduinoWait)

def readlnSerial(port):
    return str(ser.readline().decode("utf-8"))
    
def countMutations(confString):
    count = 0
    return [int(s) for s in str.split(confString) if s.isdigit()]

def getDefinition(libraryName, type):
    icData = ""
    with open(libraryName) as file: 
        for line in file:
            if (line.startswith("#version ")):
                libVersion = line.replace("#version ", "").strip()
                print("Library version: " + libVersion);
                continue
            if (line.startswith("#") or not(line.strip())):
                continue
    #        print (line.strip())
            pattern = "{\"type\": \"" + type + "\""
    #        print (" pattern:" + pattern)
            if (line.startswith(pattern)):
#                icData = json.loads("{" + line.strip() +"}")
                icData = json.loads(line.strip())
                print (" >>>> icData defined")
    return icData
    
def findQueryPins(config):
    pinDefs = config[2:]
    queryPins = []
    pins = (pinDefs.split(','))
    pinCount = 1
    for p in pins:
#        print(" p: " + p)
        if (p == 'Q' or p == 'q' or p == '-'):
            queryPins.append(pinCount)
        pinCount += 1
    return queryPins
    
def findExercisePins(icConf):
    usedExercisePins = []
    normalisedConfig = replaceChars(icConf.get("config"), "Cc")
#    print("normalisedConfig: " + normalisedConfig)
    allExercisePins = re.findall(r'\d+', normalisedConfig)
    queryPins = findQueryPins(normalisedConfig)
    for p in queryPins:
        mutation = icConf.get('M' + str(p))
#        print("Mutation: " + 'M' + str(p) + " " + mutation)
        usedExercisePins += (re.findall(r'\d+', mutation))
    uniqueList = list(set(usedExercisePins))
    uniqueList = [int(x) for x in uniqueList]
    uniqueList.sort()
#    print("ExePins: " + str(uniqueList))
    usedExercisePins = uniqueList
    return usedExercisePins
    
# "C:1,2,C,4,Q,Q,G,Q,Q,10,C,12,13,V"
def replaceChars(configStr, replaceChars):
    pins = configStr[2:].split(',')
#    print(str(pins))
    pinIndex = 0
    for rc in list(replaceChars):
        for pin in pins:
            if pins[pinIndex] == rc:
                pins[pinIndex] = str(pinIndex+1)
            pinIndex += 1
        pinIndex = 0
    return configStr[:2] + ",".join(pins)

def int2bits(value, size):
#    print("int2bits: " + str(value) + ", " + str(size))
    bitPattern = ""
    bitIndex = size - 1
    while (bitIndex >= 0):
        bit = value & 1 << bitIndex
        if bit == 0:
            bitPattern += "0"
        else:
            bitPattern += "1"
        bitIndex -= 1
    return bitPattern

def template2query(pattern, template, exercisePinList):
    bitPattern = pattern[::-1]
#    print(str(pinPatternValue) + " " + bitPattern)
    bitIndex = 0
    query = template
    for bit in bitPattern:
#        print(bit + " " + query + " " + str(bitIndex), end='')
        query = re.sub('%' + str(exercisePinList[bitIndex]) + '%', bit, query, 1)
        bitIndex += 1
#        print()
    query = re.sub(r'%\d+%', '0', query)
    return query
    
def query2result(query, pattern, queryPinList, icConf):
    bitPattern = pattern[::-1]
    normalisedQuery = replaceChars(query, 'Cc')
    expectedResultTemplate = re.sub('Q:', 'R:', normalisedQuery)
#    exercisePinList = re.findall(r'\d+', icConf.get("config"))
    exercisePinList = findExercisePins(icConf)
    
    for qp in queryPinList:
        pinDef = icConf.get('M' + str(qp))
        pinDefMod = pinDef
        epIndex = 0
        for ep in exercisePinList:
#            print("pinDefMod: " + pinDefMod)
            pinDefMod = re.sub(r'!(\d+)', r'not(\1)', pinDefMod)
            pinDefMod = re.sub(r'(\d+)', r'%\1%', pinDefMod)
            pinDefMod = re.sub(r'[%]+', r'%', pinDefMod)
            pinDefMod = re.sub('!', 'not', pinDefMod)
            pinDefMod = re.sub('not(True)', 'False', pinDefMod) # optimizers
            pinDefMod = re.sub('not(False)', 'True', pinDefMod)
            if bitPattern[epIndex] == '0':
                logicValue = 'False'
            else:
                logicValue = 'True'    
            pinDefMod = re.sub('%' + str(ep) + '%', logicValue, pinDefMod)
#            print("ep/qp: "+ ep + "/" + str(qp) + "  " + bitPattern[epIndex] + "  " + pinDef + "  " + pinDefMod)
            epIndex += 1
#        if eval(pinDefMod):
            
#        print(" pinDefMod: " + pinDefMod + "  ", end = '')
#        print(str(eval(pinDefMod)))
        if eval(pinDefMod):
            expectVal = 'H'
        else:
            expectVal = 'L'
        expectedResultTemplate = re.sub('-', expectVal, expectedResultTemplate, 1) # this assumes the query pins are 'in order'
        expectedResultTemplate = re.sub(r'%\d+%', "0", expectedResultTemplate)  # clean up unused exercise pins
    return expectedResultTemplate

#icConf = {"type": "7402", "pins": 14, "config": "C:Q,2,3,Q,5,6,G,8,9,Q,11,12,Q,V", "M1": "!(2|3)",}
#configStr = icConf["config"]

if len(sys.argv) > 1:
    ttyPort = sys.argv[1]
if len(sys.argv) > 2:
    icType = sys.argv[2]
  
print ("I.C. tester version: " + VERSION)
print ("ttyPort: " + ttyPort + " at " + str(ttySpeed) + " Baud")
print ("IC: " + icType)

icConf = getDefinition(libraryName, icType)

if (not(icConf)):
    print("ERROR: '" + icType + "' not found in library " + libraryName)
    exit()

ser = serial.Serial(ttyPort, ttySpeed, timeout=2)  # open serial port
log("> " + readlnSerial(ser).strip())

# Send IC configuration to Arduino
configStr = icConf.get("config")
log("< " + configStr)
writeSerial(ser, configStr)
time.sleep(arduinoWait)
log("> " + readlnSerial(ser).strip())


# Exercise loop init
exercisePinList = re.findall(r'[\dCc]+', configStr[2:])
usedExercisePinList = findExercisePins(icConf)
exercisePinCount = len(exercisePinList)
usedExercisePinCount = len(usedExercisePinList)
#print("Exercising " + str(exercisePinCount) + " " + str(exercisePinList))

queryPinList = findQueryPins(configStr)
queryPinCount = len(queryPinList)
#print("Querying " + str(queryPinCount) + " " + str(queryPinList))

pinInventory = exercisePinCount + queryPinCount + 2
if pinInventory != icConf.get("pins"):
    print ("ERROR: mismatch between specified pins: " + str(icConf.get("pins")) + " and found pins: " + str(pinInventory))
    exit(1)

configStrMod = re.sub(r'(\d+)', r'%\1%', replaceChars(configStr, 'Cc'))
queryStrTempl = re.sub(r'[Qq]', '-', configStrMod)
queryStrTempl = re.sub('C:', 'Q:', queryStrTempl, 1)
#print("  c: " + configStrMod + "  q: " + queryStrTempl)

exerciseCount = pow(2, usedExercisePinCount)
#exerciseCount = 3

#print(" used exe-pins: " + str(usedExercisePinCount) + " , exercise count: " + str(exerciseCount))

pinPatternValue = 0
errorCount = 0
# start the loop
for value in range (exerciseCount):
    pattern = int2bits(value, usedExercisePinCount)
    queryStr = template2query(pattern, queryStrTempl, usedExercisePinList)
    
    log("< " + queryStr)
    writeSerial(ser, (queryStr))

    actualResult   = readlnSerial(ser).strip()
    expectedResult = query2result(queryStr, pattern, queryPinList, icConf)

    log("> " + actualResult)

    if (actualResult == expectedResult):
        print (str(value + 1) + "/" + str(exerciseCount) + " " + actualResult + " Ok")
    else:
        errorCount += 1
        print("Err[" + str(errorCount) + "] '" + actualResult   + "' actual")
        print("Err[" + str(errorCount) + "] '" + expectedResult + "' expected")

resetStr = "R"

log("< " + resetStr)
writeSerial(ser, resetStr)

log("> " + readlnSerial(ser).strip())

if errorCount:
    print("testing " + icType + "; " + str(errorCount) + " errors")
else:
    print("testing " + icType + " OK.")

# find the I.C. definition in the library
# send the config to the Arduino
# loop through
## generate a bit pattern for all the exercise pins
## fill the query-template with the exercise logic levels
## fill the expected result template with exercise logic levels
## calculate the values for the query pins (convert to Python logic expression and evaluate)
## fill the expected result template with query logic levels
## send the query to the Arduino
## retrieve the response from the Arduino
## compare the expected result with the actual result
# declare result and exit
